src/ancestry_proof.rs

```diff
@@ -136,109 +135,82 @@ impl<T: Clone + PartialEq, M: Merge<Item = T>> NodeMerkleProof<T, M> {
     }
 }
 
+fn cmp_nodes<T>(a: &(u64, T, u8), b: &(u64, T, u8)) -> Ordering {
+    let (a_pos, _, a_height) = a;
+    let (b_pos, _, b_height) = b;
+    (a_height, a_pos).cmp(&(b_height, b_pos))
+}
+
 fn calculate_peak_root<
     'a,
     T: 'a + PartialEq,
     M: Merge<Item = T>,
-    // I: Iterator<Item = &'a T>
+    // I: Iterator<Item = &'a (u64, T)>,
 >(
     nodes: Vec<(u64, T)>,
     peak_pos: u64,
     // proof_iter: &mut I,
 ) -> Result<T> {
-    debug_assert!(!nodes.is_empty(), "can't be empty");
-    // (position, hash, height)
-
-    let mut queue: VecDeque<_> = nodes
+    let mut queue: VecDeque<_> = VecDeque::new();
+    for value in nodes
         .into_iter()
         .map(|(pos, item)| (pos, item, pos_height_in_tree(pos)))
-        .collect();
-
-    let mut sibs_processed_from_back = Vec::new();
+    {
+        if !queue.insert_sorted_by(value, cmp_nodes) {
```

Comment:
previously, the q was sorted by the pos first and now it's (height, pos)
my understanding is that the `nodes` will be sorted by `pos`, so this algorithm might be running in $\mathcal{O}({nˆ2})$, is that correct?

Reply:
In the worst case scenario we want to prove all the nodes. Let's say the number of nodes = n. The complexity of inserting in an ordered queue is O(log(n)). So I think the total complexity would be O(n * log(n)).

But if we want to prove just one node, we need to start from it and go up until we reach the root. Let's say the number of leaves = l. The height of the MMR is log(l). The number of nodes in the proof is at most the height of the MMR (we need at most one node per level) so the queue length is at most log(l). So the complexity should be O(log(l) * log(l))

I hope I'm not missing anything.

Reply:
> The complexity of inserting in an ordered queue is O(log(n)).

This is not true. Insertion at a random position in a VecDeque is $\mathcal{O}(n)$, because if you insert in the middle, you need to [shift](https://doc.rust-lang.org/1.80.1/src/alloc/collections/vec_deque/mod.rs.html#1868) every element on the right. If you want to guarantee logarithmic complexity, I'd suggest using a BTreeSet instead of a VecDeque.
But if the number in the proof is always bounded to a small number, that's probably fine as is. Just want to make sure we are not introducing a DoS vector in the worst case.

Reply:
Good point, you're right. It's O(n), because it's a vec-based structure. I missed that.

For our use case (ancestry proofs), yes the number of nodes in the proof should be small, but we should probably make this scalable in general. I will change it to a BTreeSet or a LinkedList.

Reply:
LinkedList would still have $\mathcal{O}(n)$ for finding the insertion position, so the right choice is BTreeSet 👍🏽

Reply:
> LinkedList would still have O ( n ) for finding the insertion position,

If it's sorted I think it should have O(log(n)). But yes, `BTreeSet` would have O(1) amortized. It's only that for small numbers it would be worse than a sorted linked list. And usually the number of nodes should be small.
BinaryHeap is also a good option.

Reply:
Or sorry, in a linked list it's not that easy to take advantage of the fact that it's sorted. You're right.

Reply:
Yes, BTreeSet/BTreeMap since to be the best option. Done.